src/openms/source/ANALYSIS/ID/TopPerc.cpp

```diff
@@ -0,0 +1,798 @@
+//  * Neither the name of any author or any participating institution
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Mathias Walzer $
+// $Authors: Mathias Walzer, Matthew The $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/config.h>
+#include <OpenMS/ANALYSIS/ID/TopPerc.h>
+
+using namespace std;
+
+namespace OpenMS
+{    
+    void TopPerc::addMSGFFeatures(vector<PeptideIdentification>& peptide_ids, StringList& feature_set)
+    {
+      feature_set.push_back("MS:1002049"); // unchanged RawScore
+      feature_set.push_back("MS:1002050"); // unchanged DeNovoScore
+      feature_set.push_back("MSGF:ScoreRatio");
+      feature_set.push_back("MSGF:Energy");
+      feature_set.push_back("MSGF:lnEValue");
+      feature_set.push_back("IsotopeError"); // unchanged IsotopeError
+      feature_set.push_back("MSGF:lnExplainedIonCurrentRatio");
+      feature_set.push_back("MSGF:lnNTermIonCurrentRatio");
+      feature_set.push_back("MSGF:lnCTermIonCurrentRatio");
+      feature_set.push_back("MSGF:lnMS2IonCurrent");
+      feature_set.push_back("MSGF:MeanErrorTop7");
+      feature_set.push_back("MSGF:sqMeanErrorTop7");
+      feature_set.push_back("MSGF:StdevErrorTop7");
+      
+      for (vector<PeptideIdentification>::iterator it = peptide_ids.begin(); it != peptide_ids.end(); ++it)
+      {
+        for (vector<PeptideHit>::iterator hit = it->getHits().begin(); hit != it->getHits().end(); ++hit)
+        {
+          // Some Hits have no NumMatchedMainIons, and MeanError, etc. values. Have to ignore them!
+          if (hit->metaValueExists("NumMatchedMainIons"))
+          {
+            // only take features from first ranked entries and only with meanerrortop7 != 0.0
+            if (hit->getMetaValue("MeanErrorTop7").toString().toDouble() != 0.0)
```

Comment:
Well, it could be if MS-GF+ output changes or the file handler messes things up. But if ```NumMatchedMainIons``` is present, so should that one. 
Otherwise, I'd have to plaster all the code with awkward checks for MetaValue existence.